Feature: remove async flag, make api async by default

DyrellC opened the pull request:
# Description of change
Update functionality to remove double api in channels. Removes the `async` flag altogether and places more reliance on `sync-client` flag instead. C bindings have been updated but wasm still has yet to be tested due to the `iota.rs` wasm `mio` issue (should be resolved soon).

ToDo:
- [x] Update wasm bindings once the `iota.rs` bug is resolved
- [x] Update any relevant docs 
- [x] Pull in Arnau's lint review

## Type of change
- Enhancement (breaking change but improves usability) 

## How the change has been tested
-  iota_streams/examples
-  iota-streams-examples 
-  c bindings example tested, only one that doesn't work is static build with client on, because corrosion doesn't accept feature flags right now

## Change checklist
- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes

DyrellC committed:
app message types updated for async

DyrellC committed:
Update fork and repeated, but functions still not working

DyrellC committed:
change fork and repeated

DyrellC committed:
Remove fork and repeat usage in keyload and user

DyrellC committed:
update examples for async

DyrellC committed:
Remove unused code, update for async-client flag

DyrellC committed:
Remove 'async' flag fully, update (not(feature = sync-client)) functionality

DyrellC committed:
update examples

DyrellC committed:
Update c bindings

DyrellC committed:
fmt

semenov-vladyslav commented on bindings/c/src/api/mod.rs:
```diff
@@ -144,7 +147,15 @@ pub extern "C" fn drop_unwrapped_messages(ms: *const UnwrappedMessages) {
+
+pub fn run_async<R>(fut: impl Future<Output=R>) -> R {
+    #[cfg(feature = "sync-client")]
+    let ret = tokio::runtime::Runtime::new().unwrap().block_on(fut);
```
Is it OK to always create a new Runtime?

DyrellC replied:
There are 2 options really: 
1. Create a new runtime each time we run a function 
2. Wrap author instance with a stored runtime inside it and spawn blocking from it every time we need to use a call to send/receive 

It is necessary to have a tokio runtime if we are using the iota.rs client, that much is certain

semenov-vladyslav commented on iota-streams-app-channels/src/api/tangle/user.rs:
```diff
@@ -183,363 +183,45 @@ impl<Trans> User<Trans> {
-    /// # Arguments
-    /// * `link` - Address of message to act as root of previous message fetching
-    pub fn fetch_prev_msg(&mut self, link: &Address) -> Result<UnwrappedMessage> {
+    async fn transport_recv(&mut self, link: &Address) -> Result<Message> {
```
Git diff is super weird, it decided that it's a proper place to add these two lines between these big chunks of removed code.

semenov-vladyslav commented on iota-streams-app/src/message/content.rs:
```diff
@@ -9,31 +13,35 @@ use iota_streams_ddml::{
+#[async_trait]
 pub trait ContentSizeof<F> {
-    fn sizeof<'c>(&self, ctx: &'c mut sizeof::Context<F>) -> Result<&'c mut sizeof::Context<F>>;
+    async fn sizeof<'c>(&self, ctx: &'c mut sizeof::Context<F>) -> Result<&'c mut sizeof::Context<F>>;
```
Why exactly this should be `async`? It's supposed to be a pure function.

DyrellC replied:
These functions are going to need to be async to accommodate the DID functionality since everything on that end is asynchronous. It's either `block_on`'s or an async trait.

semenov-vladyslav commented on iota-streams-app/src/message/prepared.rs:
```diff
@@ -18,14 +21,14 @@ use iota_streams_ddml::{
 /// Message context prepared for wrapping.
 pub struct PreparedMessage<'a, F, Link: Default, Store: 'a, Content> {
-    store: Ref<'a, Store>,
+    store: &'a Arc<Mutex<Store>>,
```
Shall we abstract `Arc<Mutex<..>>` away somehow? Or maybe even add constraints to `LinkStore` trait so thread-safety is guaranteed?

DyrellC replied:
Seems a reasonable thing to do. Where were you thinking? Just in `message/mod.rs`?

semenov-vladyslav commented on iota-streams-app-channels/src/message/announce.rs:
```diff
@@ -97,13 +103,15 @@ impl<F> Default for ContentUnwrap<F> {
-        &mut self,
-        _store: &Store,
+    async fn unwrap<'c, IS: io::IStream>(
+        &'c mut self,
```
I'm not sure it has to be the same life-time for `&mut self` and `store: &Store` as for `ctx` and result type. But if it works then fine.

DyrellC replied:
There was a compiler complaint at some point that was pushing for me to set the lifetime for self as well, but it looks like it's not complaining about it anymore so i'll remove it for now, because I agree it shouldn't need the same constraint

semenov-vladyslav commented on iota-streams-app-channels/src/message/keyload.rs:
```diff
@@ -93,87 +95,107 @@ pub struct ContentWrap<'a, F, Link: HasLink, Keys> {
+        ctx.join(store, self.link)?.absorb(&self.nonce)?;
+
+        // fork into new context in order to hash Identifiers
+        let saved_fork = ctx.spongos.fork();
```
What is the issue with `fork`? Why workaround is needed?

DyrellC replied:
Fork won't let me insert async functions, neither will repeated, it creates lifetime conflicts that I was unable to find a workaround for.

semenov-vladyslav commented on iota-streams-app/src/message/prepared.rs:
```diff
@@ -43,26 +46,26 @@ impl<'a, F, Link: Default, Store: 'a, Content> PreparedMessage<'a, F, Link, Stor
+    Store: 'a + LinkStore<F, <Link as HasLink>::Rel> + Send + Sync,
     HDF<Link>: ContentWrap<F, Store>,
-    Content: ContentWrap<F, Store>,
+    Content: ContentWrap<F, Store> + Send + Sync,
```
Maybe add `Send + Sync` constraints to all these traits `ContentWrap`, `LinkStore`, `HasLink`, etc.?

DyrellC replied:
Yeah the thought crossed my mind. Would save on all these extra trait definitions everywhere

semenov-vladyslav commented on iota-streams-app/src/transport/bucket.rs:
```diff
@@ -37,6 +41,7 @@ where
     }
 }
 
+#[cfg(feature = "sync-client")]
```
Why constraint here? `TransportOptions` can be implemented for `BucketTransport` anyway, can't it?

semenov-vladyslav commented on iota-streams-app/src/transport/mod.rs:
```diff
@@ -154,77 +169,97 @@ impl<Link: Debug + Display, Msg, Tsp: Transport<Link, Msg>> Transport<Link, Msg>
-impl<Tsp: TransportOptions> TransportOptions for Arc<AtomicRefCell<Tsp>> {
+#[cfg(not(feature = "sync-client"))]
+#[async_trait]
+impl<Tsp: TransportOptions + Send + Sync> TransportOptions for Arc<Mutex<Tsp>>
```
Is `Mutex` better than `AtomicRefCell`?

DyrellC replied:
In hindsight perhaps not, originally it was the mutex from futures since we were already loading it in, but now it uses spin (since it's lightweight and `no_std` ready) so it's worth discussing the options

semenov-vladyslav commented on iota-streams-app/src/transport/mod.rs:
```diff
@@ -51,10 +53,22 @@ pub trait TransportOptions {
 /// Parametrized by the type of message links.
 /// Message link is used to identify/locate a message (eg. like URL for HTTP).
-#[cfg(not(feature = "async"))]
+#[cfg(feature = "sync-client")]
```
Once again, maybe just get rid off of `sync-client` feature and this version (non-async) of `Transport`?

semenov-vladyslav commented on iota-streams-app/src/transport/tangle/client.rs:
```diff
@@ -170,19 +171,19 @@ pub async fn async_get_link_details(client: &iota_client::Client, link: &TangleA
 
 /// Synchronised - Retrieve a message from the tangle using a node client
-#[cfg(not(feature = "async"))]
+#[cfg(feature = "sync-client")]
```
No `sync-client` feature, no `block_on`s, no problems. Profit

DyrellC replied:
Right? Think it's better to just remove it entirely as you suggested. I was going to suggest the same thing for now

semenov-vladyslav commented on iota-streams-ddml/src/command/wrap/fork.rs:
```diff
@@ -9,11 +9,11 @@ use iota_streams_core::sponge::prp::PRP;
 impl<C, F: PRP, OS: io::OStream> Fork<C> for Context<F, OS>
 where
-    C: for<'a> FnMut(&'a mut Self) -> Result<&'a mut Self>,
+    C: for<'a> FnMut(&'a mut Self) -> Result<&'a mut Self> + Send,
```
Why is this `fork` not used in wrap/unwrap context? Constraint on `C` should be something like `for<'a> async FnMut(..) ..` or `async` can be made explicit in the result type: `for<'a> FnMut(..) -> Future<Output = Result<..>>`

DyrellC replied:
I tried the second option and ran into a bunch of lifetime headaches (especially in `api/user.rs`)

DyrellC replied:
Saving the fork and repeated for a different pr

semenov-vladyslav commented on iota-streams-ddml/src/command/wrap/fork.rs:
```diff
@@ -9,11 +9,11 @@ use iota_streams_core::sponge::prp::PRP;
-    C: for<'a> FnMut(&'a mut Self) -> Result<&'a mut Self>,
+    C: for<'a> FnMut(&'a mut Self) -> Result<&'a mut Self> + Send,
 {
     fn fork(&mut self, mut cont: C) -> Result<&mut Self> {
```
`fork` method should definitely be `async` if we allow async Wrap/Unwrap traits.

semenov-vladyslav reviewed (commented):
Looks good but there are a few places that need clarification.

DyrellC commented on bindings/c/Cargo.toml:
```diff
@@ -5,9 +5,9 @@ authors = ["Vlad Semenov <vlad.semenov@iota.org>", "Dyrell Chapman <dyrell.chapm
-std = ["iota-streams/std"]
-sync-client = ["iota-streams/sync-client"]
+default = ["std", "iota-streams/async-client"]
+std = ["iota-streams/std", "iota-streams/async-client"]
```
Ah yes, definitely the better way to be doing that

DyrellC commented on bindings/c/src/api/auth.rs:
```diff
@@ -306,30 +317,49 @@ pub unsafe extern "C" fn auth_send_signed_packet(
 /// Process a Signed packet message
 #[no_mangle]
-pub unsafe extern "C" fn auth_receive_signed_packet(
+pub async unsafe extern "C" fn auth_receive_signed_packet(
```
Nah it's a remnant from experimenting with different approaches (specifically from when there wasn't a utility function for running asynchronous code so each function was doing it individually with awaits etc).

DyrellC commented on bindings/c/src/api/auth.rs:
```diff
@@ -306,30 +317,49 @@ pub unsafe extern "C" fn auth_send_signed_packet(
+    if let Some(r) = r.as_mut() {
+        if let Some(user) = user.as_mut() {
+            if let Some(link) = link.as_ref() {
+                return user.receive_signed_packet(link).await.map_or(Err::OperationFailed, |signed_payloads| {
```
It should indeed be `run_async`

DyrellC committed:
Remove lifetime for self in unwrap

DyrellC committed:
fix typos and artifacts in c bindings

DyrellC commented on iota-streams-app/Cargo.toml:
```diff
@@ -14,13 +14,12 @@ default = ["std", "tangle", "sync-client"]
 sync-client = ["num_cpus", "iota-client/sync", "tangle", "std"]
-async-client = ["num_cpus", "iota-client/default", "tangle", "async", "std"]
-wasm-client = ["iota-client/wasm", "chrono/wasmbind", "tangle", "async", "std"]
+async-client = ["num_cpus", "iota-client/default", "tangle", "std"]
```
I was going to ask the same thing to everyone, I think it's an unnecessary feature at this time unless we plan to separate the client functionality to actually allow for sending the messages asynchronously, but since that's not present at the time it's just superfluous and bloats the code base.

sascha1337 replied:
Please go for it :) It would make the debugging and everything so much easier, with that less complexity and less feature deps chaining. It really confused not only me :P

DyrellC committed:
Consolidate Send + Sync traits

DyrellC committed:
Remove 'sync-client' and 'async-client' flags

DyrellC committed:
comment out main_client() in examples

DyrellC committed:
use spin mutex

DyrellC committed:
c bindings feature update

DyrellC committed:
c bindings doc

DyrellC committed:
More feature flag updates

DyrellC committed:
update c bindings

DyrellC committed:
update wasm bindings

DyrellC committed:
clippy/fmt

arnauorriols commented:
> Enhancement (a non-breaking change which adds functionality)
How come you consider it a non-breaking change? I'd say it's a pretty significant breaking change...

kwek20 commented:
haha and also mainly removes functionality, not adds it :D

DyrellC commented:
> > Enhancement (a non-breaking change which adds functionality)
> 
> How come you consider it a non-breaking change? I'd say it's a pretty significant breaking change...

Oh that's definitely a typo. I meant to remove the `non-breaking change` part of that but obviously did not

DyrellC committed:
Fix rebase artifacts

DyrellC committed:
Use global runtime for c bindings

DyrellC commented on iota-streams-app/src/transport/mod.rs:
```diff
@@ -101,140 +67,172 @@ where
+// Msg: Send + Sync,
+// {
+// Send a message with default options.
+// async fn send_message(&mut self, msg: &Msg) -> Result<()>;
```
Definitely dead code remnants. I'm removing them.

DyrellC committed:
formatting

DyrellC committed:
Move ToString out of feature import

DyrellC committed:
Update single depth logic for async

DyrellC committed:
Remove unnecessary Send + Sync

Commit:
Remove all the Sync and Send trait bounds

Trait bounds should generally be pushed to the latest moment when they are actually needed.
That is, the functions that manipulate the data and that actually require the bound.
It is usually avoided (to the reasonable extend) to put trait bounds in the type parameters
of data types and trait declarations, as they unleash an unergonomic "trait bound hell" and are usually overly restrictive.

Some relevant references:
- [API Guidelines]
- [discussion API Guidelines]
- [flat2 case]
- [Haskell Programming Guidelines]

In this case, the trait bounds "Sync" and "Send" are already enforced (if necessary) when the async runtime actually schedule the Futures,
and it is at this point (if actually enforced by the runtime) that all the types and intermediate functions that manipulate those types are required
to enforce these markers; However, if at any point of the rabbit hole a type bound is placed on the data, hell breaks loose and all types that touch
the bounded type get contaminated. [Implied bounds RFC] might mitigate the issue, but for the time being, this convention is gold one of those that
is "all or nothing", because of the "viral" effect.

[API Guidelines]: https://rust-lang.github.io/api-guidelines/future-proofing.html#data-structures-do-not-duplicate-derived-trait-bounds-c-struct-bounds
[discussion API Guidelines]: https://github.com/rust-lang/api-guidelines/issues/6
[Haskell Programming Guidelines]: https://wiki.haskell.org/Programming_guidelines#Types
[flate2 case]: https://github.com/rust-lang/flate2-rs/issues/88
[Implied bounds RFC]: https://github.com/rust-lang/rust/issues/44491

arnauorriols reviewed (changes requested):
This is a first batch of comments. I have a couple of stashed issues that I want to make sure are feasible to improve before proposing. 

Overall, this PR is a very significant improvement in terms of design, amazing job!

DyrellC commented on iota-streams-core/Cargo.toml:
```diff
@@ -26,6 +27,9 @@ rand = { version = "0.7", default-features = false, features = ["wasm-bindgen"]}
 anyhow = { version = "1.0.34", default-features = false, optional = false }
+async-trait = { version = "0.1", optional = false }
+futures = { version = "0.3.8", default-features = false, features = ["executor"] }
+spin = { version = "0.9.2", default-features = false, features = ["mutex", "spin_mutex"] }
```
Caught in the fallacy that it would be a reasonable `#[no-std]` compatible replacement, thanks for the reading material. I was hoping to use just one implementation if possible but i'll work on replacing that.

DyrellC commented on iota-streams-core/Cargo.toml:
```diff
@@ -26,6 +27,9 @@ rand = { version = "0.7", default-features = false, features = ["wasm-bindgen"]}
 hex = { version = "0.4", default-features = false, optional = false, features = ["alloc"] }
 anyhow = { version = "1.0.34", default-features = false, optional = false }
+async-trait = { version = "0.1", optional = false }
+futures = { version = "0.3.8", default-features = false, features = ["executor"] }
```
For `async-trait` it was so it could be accessed from ddml, `futures` was just sort of along for the ride. I'm moving `futures` back but i'm going to keep `async-trait` where it is to tackle to the `repeated/fork` issue in a different PR.

DyrellC commented on iota-streams-core/Cargo.toml:
```diff
@@ -12,6 +12,7 @@ description = "A rust implementation of the IOTA Streams core utils"
 default = ["std"]
 # enable std
 std = ["rand/std", "digest/std", "hex/std"]
+no-std = []
```
Yup, removing

DyrellC commented on iota-streams-core/src/sponge/prp/prp.rs:
```diff
@@ -7,7 +7,7 @@ use crate::prelude::generic_array::{
 /// Actually, it may be non-bijective as the inverse transform is not used in sponge construction.
 #[allow(clippy::upper_case_acronyms)]
-pub trait PRP: Sized + Default + Clone //+ From<Vec<u8>> + Into<Vec<u8>>
+pub trait PRP: Sized + Default + Clone + Send + Sync //+ From<Vec<u8>> + Into<Vec<u8>>
```
I'm ok with going the `(?Send)` route, I just hope it doesn't make some new headaches in the DID realm. I don't suspect it will but you never know dealing with async stuff in rust :man_shrugging:

DyrellC replied:
I'll get that merged in after I've wrapped up a couple other changes here

DyrellC commented on iota-streams-ddml/src/command/wrap/fork.rs:
```diff
@@ -9,11 +9,11 @@ use iota_streams_core::sponge::prp::PRP;
     fn fork(&mut self, mut cont: C) -> Result<&mut Self> {
         let saved_fork = self.spongos.fork();
-        cont(self)?;
+        cont(&mut *self)?;
```
Remnant from async fork and repeated experiments, i'll revert it back

DyrellC commented on iota-streams-app/src/message/wrapped.rs:
```diff
@@ -21,7 +21,7 @@ impl<F: PRP, Link: HasLink> WrapState<F, Link> {
     /// Save link for the current wrapped message and associated info into the store.
```
Not sure I follow on this one? What are you referring to specifically?

arnauorriols replied:
Oh, I understand now. it was regarding the `accociated` instead of `associated` somehow I saw that in your PR you were introducing the typo to an already correctly written word, but now I believe that was a git conflict resolution gone wrong. Nevermind 😅

DyrellC committed:
move futures, update mutex/mutexguard

DyrellC committed:
Remove unnecessary borrow/deref in fork

DyrellC committed:
Comment for clarity

DyrellC commented on iota-streams-app/src/transport/mod.rs:
```diff
@@ -89,152 +55,82 @@ where
 }
 
-impl<Tsp: TransportOptions> TransportOptions for Rc<RefCell<Tsp>> {
+pub type SharedTransport<T> = Rc<RefCell<T>>;
```
Yeah it hasn't been used anywhere, I think it's reasonable to remove it

DyrellC commented on iota-streams-app/src/transport/mod.rs:
```diff
@@ -41,40 +27,20 @@ where
 }
 
+#[async_trait(?Send)]
 pub trait TransportOptions {
```
Probably not really, I'll go ahead and remove it provided compilation doesn't give me any grief doing so.

DyrellC replied:
It is not, changed now

DyrellC committed:
TransportOptions no longer async

DyrellC commented on iota-streams-app/src/message/content.rs:
```diff
@@ -9,31 +13,35 @@ use iota_streams_ddml::{
+    async fn unwrap<'c, IS: io::IStream>(
         &mut self,
-        store: &Store,
+        store: &'c Store,
```
Keyload was giving grief at some point and then it just got left the way it was. Looks like the compiler isn't spitting out the same errors anymore, so I'll revert it back

DyrellC committed:
no lifetime on unwrap store

DyrellC committed:
Merge branch 'feat/remove-async-flag' into remove-send-sync-bounds

DyrellC committed:
Merge pull request #2 from arnauorriols/remove-send-sync-bounds

Remove all the Sync and Send trait bounds

DyrellC committed:
Link store to rwlock instead of mutex

DyrellC commented on iota-streams-app/src/message/preparsed.rs:
```diff
@@ -21,17 +24,18 @@ impl<'a, F, Link: Default + Clone> PreparsedMessage<'a, F, Link> {
+    pub async fn unwrap<Store, Content>(
         mut self,
-        store: &Store,
+        store: &MutexGuard<'_, Store>,
```
I like that, i'll have to make a few extra little changes to accommodate it though

DyrellC commented on iota-streams-app-channels/src/message/keyload.rs:
```diff
@@ -99,84 +104,105 @@ where
+
+        // fork into new context in order to hash Identifiers
+        {
+            ctx.absorb(repeated_keys)?;
```
I was using it for my own reference but they're pretty pointless now. Removing

DyrellC commented on iota-streams-app-channels/src/api/user.rs:
```diff
@@ -1043,34 +1058,36 @@ where
+        ctx.absorb(repeated_keys)?;
+        for key in keys {
+            let (id, cursor) = key;
+            let ctx = id.clone().wrap(_store.borrow(), ctx.borrow_mut()).await?;
```
Yeah that underscores been there since the ancient days lol I'll fix that up

DyrellC commented on iota-streams-app-channels/src/api/tangle/author.rs:
```diff
@@ -322,11 +158,14 @@ impl<Trans: Transport + Clone> Author<Trans> {
+        let ann = author.user.user.announce().await?;
+        #[cfg(feature = "sync-client")]
+        let retrieved: Message = author.user.transport.recv_message(announcement)?;
+        #[cfg(not(feature = "sync-client"))]
```
The merge must have left that in there I definitely remember correcting this at one point.

DyrellC replied:
hurray for rebase right?

arnauorriols replied:
I figured as much 😅

DyrellC commented on bindings/c/Cargo.toml:
```diff
@@ -16,5 +16,7 @@ crate-type = ["staticlib", "cdylib", "rlib"]
 
 [dependencies]
 iota-streams = {path = "../../", default-features = false, features = ["tangle"]}
+tokio = {version = "1.9.0", features = ["rt", "rt-multi-thread"]}
```
We need tokio no matter which implementation we use. It's part of the global runtime we have to set up for the `run_async` wrapper function

arnauorriols replied:
Yeah, you are right. I was thinking in terms of "we only need async for the transport client" but as long as even DDML is `async`, `tokio` goes with us to the bone...

DyrellC committed:
General cleanup

DyrellC committed:
fmt

DyrellC committed:
Add no-std feature again

DyrellC committed:
lifetime in user

DyrellC committed:
restrict spin in prelude

DyrellC committed:
Update transport options in wasm

Commit:
Avoid using `Arc<RwLock<T>>` for `LinkStore`

It's usage was forced due to lifetime missmanagement and overrestrictive Send + Sync bounds.

Commit:
Remove Arc<Mutex<T>> synchronization from wasm bindings

Commit:
Remove synchronization from C bindings

Commit:
Remove synchronization from main examples

Commit:
Feature gate sync Transport implementation

2 new features added, to be able to chose between spin-flavoured or parking-lot-flavoured Mutex

Commit:
Fix option SYNC_CLIENT => IOTA_CLIENT in CMakeLists

Commit:
Remove unused feature "serde-serialize" from wasm-bindgen dependency

Commit:
Re-wrap the inner Client of the wasm types although with `Rc<RefCell<T>>` instead of `Arc<Mutex<T>>`

Commit:
Fix get_link_details for `Rc<RefCell<Transport>>`

Commit:
Reintroduce `TransportWrap` name in C bindings

arnauorriols commented:
So, I've been experimenting with my couple of pending issues, and have pushed DyrellC/streams#3 

In summary, I had the suspicion that the interior mutability (and now synchronized mutation) was at least not a necessity, and in my opinion not justified given the nature of the library. I've managed to remove the synchronization both in the `LinkStore` and in the bindings. Synchronization has a significant impact in performance and embedded compatibility (for example, espidf target, while implementing std, don't support any locking mechanism), therefore I think it's worth its removal if possible.

DyrellC commented:
> support

I agree, it seems excessive to be accommodating for a use case that will never be. The only instance of asynchronous functionality that should be accommodated imo is the ability to asynchronously transport the message without waiting on the response but there are other accomodations necessary before we go down that path so we'll save that for the future.

DyrellC committed:
Merge pull request #3 from arnauorriols/remove-synchronization

Remove synchronization
